### PR TITLE
[BKY-5641] Breakout COSE

### PR DESCRIPTION
**Pull request by tahardi**
## Describe your changes
Moved CoseSign1 code from `internal/nitrite.go.Verify()` into `internal/cose.go`. Specifically, code for creating a CoseSign1 structure from bytes, verifying required fields, and checking the signature.

## Related Jira cards
[break out cose](https://blocky.atlassian.net/browse/BKY-5641?atlOrigin=eyJpIjoiZGI4YWI2YWRjOTRlNDZjNWI4NzU2Y2VmNWYzNTJjYjUiLCJwIjoiaiJ9)

## Notes for reviewers
I made a slight update to `internal/document.go` to match the format introduced in `internal/cose.go`. Namely, I removed `document.Verify()` in lieu of a `MakeDocumentFromBytes()` function which handles unmarshaling and basic field checks.

## Checklist before requesting a review

If any of these checks are missing, please provide an explanation.

- [x] I have updated README files, if applicable.
- [x] I have run `make pre-pr` and all checks have passed.
- [x] This PR is small. If not, justify below.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -32,7 +32,7 @@ test-main: tidy
 
 
 .PHONY: test
-test: test-unit test-main
+test: test-unit test-integration test-main
```
Comment (tahardi) [inline Makefile:35]: Should I have added this to `make pre-pr` instead? Either way, it's easy to forget about integration testing right now since it is not in `make test` or `make pre-pr`

File: internal/cose.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+type SigningAlgorithm int
+
+const (
+	BadOrMissingAlgorithm SigningAlgorithm = iota
+	ES384
+)
```
Comment (tahardi) [inline internal/cose.go]: While COSE supports more signing algorithms, AWS Nitro only uses ES384. So, I didn't bother enumerating the rest. Should be easy enough to add additional ones if Nitro moves to a new signing algo

File: internal/cose.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+// https://datatracker.ietf.org/doc/html/rfc8152#section-4.2
+type CoseSign1 struct {
+	_ struct{} `cbor:",toarray"`
+
+	Protected   []byte
+	Unprotected cbor.RawMessage
+	Payload     []byte
+	Signature   []byte
+}
```
Comment (tahardi) [inline internal/cose.go:28]: Previously, this was referred to as `CosePayload` but it is in fact a `CoseSign1` structure as defined in the RFC I linked---this is format in which Nitro returns signed attestations.

File: internal/cose.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+type CoseHeader struct {
+	Alg interface{} `cbor:"1,keyasint,omitempty" json:"alg,omitempty"`
+}
+
+func MakeCoseSign1FromBytes(bytes []byte) (CoseSign1, error) {
```
Comment (tahardi) [inline internal/cose.go]: Where "bytes" are a base64 decoded attestation string. Handles unmarshaling into cose sign 1 struct and checking that required fields exist

File: internal/cose.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	return coseSign1, nil
+}
+
+func (c CoseSign1) CheckSignature(
+	publicKey *ecdsa.PublicKey,
+) ([]byte, error) {
+	signingAlg, err := c.GetSigningAlgorithm()
+	if err != nil {
+		return nil, fmt.Errorf("extracting signing algorithm: %w", err)
+	}
```
Comment (tahardi) [inline internal/cose.go]: One change I made was to extract the signing algo from the CoseSign1 header and ensure that it matches the signing algorithm of the public key passed in (this is enforced by the switch statement on L84)

File: internal/cose.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+		Context     string
+		Protected   []byte
+		ExternalAAD []byte
+		Payload     []byte
+	}{
+		Context:     "Signature1",
+		Protected:   c.Protected,
+		ExternalAAD: []byte{},
+		Payload:     c.Payload,
+	}
```
Comment (tahardi) [inline internal/cose.go:81]: Made `coseSignature` anonymous since it is only used here. Additionally, added link to the section of RFC that defines this structure and how to use it to verify COSE signature

File: internal/cose.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+	sigStructBytes, err := cbor.Marshal(&sigStruct)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling cose signature struct: %w", err)
+	}
+
+	var sigStructHash []byte
+	switch signingAlg {
+	case ES384:
+		if publicKey.Curve.Params().Name != "P-384" {
```
Comment (tahardi) [inline internal/cose.go:91]: Ensure the signing algo listed in CoseSign1 matches the passed in public key. The old `checkECDSASignature` code allowed any public key with P-224, 256, 384, or 512, but the AWS Nitro docs states that only 384 is used.

File: internal/document.go [hunk 8]
```diff
@@ -20,6 +22,25 @@ type Document struct {
 	Nonce     []byte `cbor:"nonce" json:"nonce,omitempty"`
 }
 
+func MakeDocumentFromBytes(bytes []byte) (Document, error) {
```
Comment (tahardi) [inline internal/document.go]: Updated document to be closer to `CoseSign1` where we have a make-from-bytes function that is responsible for unmarshaling and verifying the correctness of fields---where correctness simply means required fields exist and that things are not too short/long.

File: internal/nitrite.go [hunk 9]
```diff
@@ -97,20 +36,11 @@ func WithTime(t time.Time) VerificationTimeFunc {
...
-// verification has passed, certificate verification has failed (expired, not
-// trusted, etc.), signature is OK or signature is not OK. If either signature
-// is not OK or certificate can't be verified, both Result and error will be
-// set! You can use the SignatureOK field from the result to distinguish
-// errors.
+// The process for verifying Nitro attestations is documented here:
+// https://github.com/aws/aws-nitro-enclaves-nsm-api/blob/4b851f3006c6fa98f23dcffb2cba03b39de9b8af/docs/attestation_process.md
+//
+// Revocation checks are NOT performed and you should check for revoked
+// certificates by looking at the `Certificates` field in the `Result`.
```
Comment (tahardi) [inline internal/nitrite.go:43]: This comment was out of date. Removed most of the text and instead permalinked to the nitro attestation verification process docs.

The old comment did not that revocation checks are not performed and I thought that was an important comment to leave in.

File: internal/result.go [hunk 10]
```diff
@@ -21,10 +21,6 @@ type Result struct {
 	// Signature section from the COSE Sign1 payload.
 	Signature []byte `json:"signature,omitempty"`
 
-	// SignatureOK designates if the signature was OK (but certificate could be
-	// invalid, not trusted, expired, etc.)
-	SignatureOK bool `json:"signature_ok"`
```
Comment (tahardi) [inline internal/result.go:26]: This was needed when it was possible to return a valid Result object with an invalid signature and error. The updated behavior of `nitrite.Verify()`, however, does not allow this. Thus, `SignatureOK` is not required---if you do not get an error back then the implication is that the signature is ok

File: internal/result.go [hunk 11]
```diff
@@ -21,10 +21,6 @@ type Result struct {
...
-
 	// COSESign1 contains the COSE Signature Structure which was used to
 	// calculate the `Signature`.
 	COSESign1 []byte `json:"cose_sign1,omitempty"`
```
Comment (tahardi) [inline internal/result.go]: Looking at `Result`, are any of these variables even used? From what I can tell, the only useful information is the Document and certificate chain. Could we remove the rest and add back in future if we actually find the need for it?

File: internal/nitrite.go [hunk 12]
```diff
@@ -196,83 +69,27 @@ func Verify(
...
 
-	signatureOk := checkECDSASignature(
-		certificates[0].PublicKey.(*ecdsa.PublicKey),
-		sigStruct,
-		cose.Signature,
-	)
+	}
+	if len(certificates) < 1 {
+		return nil, fmt.Errorf("certificates chain is empty")
+	}
```
Comment (tahardi) [inline internal/nitrite.go]: Do I need this check or can I assume that if `document.CheckCertificates` returned without error then this is populated? I thought that was a fair assumption to make in Go

File: internal/cose.go (hunk 4 above)
Comment (dlm) [reply internal/cose.go]: Be careful with using "bytes" as a variable.  It is a package name in the go standard library see: https://pkg.go.dev/bytes.  (I have been bit a few times by shadowing the package a few times recently.)

File: internal/nitrite.go (hunk 12 above)
Comment (dlm) [reply internal/nitrite.go]: In some cases.  In general, docs provide guidance and how you process the data can also help with some ideas.  So for example, if you don't know the length and are iterating to see if an item exists, it doesn't matter.  If you don't know the length, it is unspecified, the operation can succeed and produce an empty slice, it isn't a bad idea to check, else you can be out of bounds.  

Here, since the code is part of the library, we get to choose.  In the current version, we must have at least one item in the slice, so if you think that is always the case, you can document it and make the assumption in functions that call it.  Alternatively, if you don't want to make that grantee you can leave it unspecified and update to something like:
```
switch {
    case err != nil:
      		return nil, fmt.Errorf("checking document certificates: %w", err)
    case len(certificates) == 0:
                return nil, fmt.Errorf("certificates chain is empty")    
}
````

Personally, I would do the check and not make the assumption because the assumption doesn't simplify the code much while additional assumptions can make maintenance more cumbersome.  (Now this is internal to the package so it may not matter that much.)

**Review by dlm — COMMENTED** [review]
I was just popping over to see how things are going on nitrite.  Didn't do a deep dive (but can if you like).  Added a few comments while I was in here.

File: internal/result.go (hunk 11 above)
Comment (mwittie) [reply internal/result.go]: Bear and I discussed and we'll remove `Result` and just return `Document`

File: internal/document.go (hunk 8 above)
Comment (mwittie) [reply internal/document.go]: My recommendation would be to convert `MakeDocumentFromBytes` (here and in `CoseSign1`) to make `Documnet` and `CoseSign1` to conform to the https://pkg.go.dev/encoding#BinaryMarshaler interface. You can do that by implementing `MarshalBinary` and `UnmarshalBinary` like:
```go
func (doc *Document) UnmarshalBinary(data []byte) error {
	return cbor.Unmarshal(data, doc)
}
```

Doing that could let you use `doc.MarshalBinary` in tests. You could also have then a `Verify` function on `Document` that would do all the verification. I'm a little concerned about `internal.Verify` needing to know a lot about how to sequence the different document verification steps - it would be easier to have them wrapped inside `Verify`. You could then implement a `Document.Certificates` functions to get the certificates for `coseSign1.CheckSignature`, but also a user getting a `Document` back (instead of a `Result`) could call `Document.Certificates`.

File: internal/cose.go (hunk 4 above)
Comment (mwittie) [reply internal/cose.go]: I suggest we change this to `CoseSign1.UnmarshalBinary` - see my comment in document.go for a discussion

File: internal/cose.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+		return CoseSign1{}, fmt.Errorf("missing cose payload")
+	}
+
+	if len(coseSign1.Signature) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose signature")
+	}
+	return coseSign1, nil
+}
+
+func (c CoseSign1) CheckSignature(
```
Comment (mwittie) [inline internal/cose.go]: My suggestion would be to change this to `Verify` to match `Document`

File: internal/document_test.go [hunk 14]
```diff
@@ -29,16 +29,89 @@ func initDocuments(t *testing.T) (
...
+		// when
+		_, err = internal.MakeDocumentFromBytes(docBytes)
+
+		// then
+		assert.ErrorContains(t, err, "checking mandatory fields")
+	})
+
+	t.Run("checking optional fields", func(t *testing.T) {
+		// given
+		attestation, err := base64.StdEncoding.DecodeString(internal.NitroAttestationB64)
```
Comment (mwittie) [inline internal/document_test.go]: long line

File: internal/document_test.go [hunk 15]
```diff
@@ -29,16 +29,89 @@ func initDocuments(t *testing.T) (
...
+		require.NoError(t, err)
+
+		coseSign1, err := internal.MakeCoseSign1FromBytes(attestation)
+		require.NoError(t, err)
+
+		doc := internal.Document{}
```
Comment (mwittie) [inline internal/document_test.go]: I think it makes sense in `happy path` tests to use the `testdata` attestations. In the error tests it would be clearer to build a `doc` from scratch and then user `doc.MarshalBinary` to pass to see if it can be `UnmarshallBinary` (if that function has any checking).

File: internal/cose_test.go [hunk 16]
```diff
@@ -0,0 +1,347 @@
+package internal_test
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/base64"
+	"testing"
+
+	"github.com/blocky/nitrite/internal"
```
Comment (mwittie) [inline internal/cose_test.go]: Could you separate blocky imports? Here and in `document_test.go`

File: internal/cose.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+				"cose signing alg '%v' does not match public key signing alg '%s'",
+				signingAlg,
+				publicKey.Curve.Params().Name,
+			)
+		}
+
+		h := sha512.Sum384(sigStructBytes)
+		sigStructHash = h[:]
+	default:
+		return nil, fmt.Errorf("unsupported signing alg '%v'", signingAlg)
```
Comment (mwittie) [inline internal/cose.go]: Does this case need a test?

File: Makefile (hunk 1 above)
Comment (mwittie) [reply Makefile:35]: Either way - there is a card to replace testing of the executable, so that we can use some go tooling for it.

**Review by mwittie — APPROVED** [review]
Looks good. I made some suggestions to use `UnmarshalBinary` and `Verify` in `CoseSign1` and `Document` to simplify what `internal.Verify` needs to know about the steps needed to verify things.

File: internal/result.go (hunk 11 above)
Comment (tahardi) [reply internal/result.go]: Removed

File: internal/nitrite.go (hunk 12 above)
Comment (tahardi) [reply internal/nitrite.go]: Thanks for weighing in. I left the check

File: internal/document_test.go (hunk 14 above)
Comment (tahardi) [reply internal/document_test.go]: outdated code no longer exists

File: internal/document_test.go (hunk 15 above)
Comment (tahardi) [reply internal/document_test.go]: Setup is much simpler now with UnmarshalBinary

File: internal/document.go (hunk 8 above)
Comment (tahardi) [reply internal/document.go]: Both Document and CoseSign1 implement `UnmarshalBinary` now. Removed the make functions and added verification functions

File: internal/cose.go (hunk 17 above)
Comment (tahardi) [reply internal/cose.go]: It's not actually reachable under the current implementation as the `GetSigningAlgorithm()` function only supports `ES384` so this type of "unsupported signing alg" gets caught above. I still need a default case for the switch statement, however, and that case would be an unsupported signing algo.

File: internal/cose_test.go (hunk 16 above)
Comment (tahardi) [reply internal/cose_test.go]: Ah, goland was only sorting "bky.sh" imports. Added `github.com/blocky/` to the ruleset as well

File: internal/document_test.go [hunk 18]
```diff
@@ -357,7 +399,7 @@ func TestDocument_CheckOptionalFields(t *testing.T) {
 		doc.UserData = make([]byte, 1025)
```
Comment (mwittie) [inline internal/document_test.go]:
```suggestion
		doc.UserData = make([]byte, internal.MaxUserDataLen+1)
```

File: internal/document_test.go [hunk 19]
```diff
@@ -369,18 +411,17 @@ func TestDocument_CheckOptionalFields(t *testing.T) {
 		doc.Nonce = make([]byte, 1025)
```
Comment (mwittie) [inline internal/document_test.go]:
```suggestion
		doc.Nonce = make([]byte, internal.MaxNonceLen+1)
```

File: internal/nitrite.go [hunk 20]
```diff
@@ -97,182 +36,53 @@ func WithTime(t time.Time) VerificationTimeFunc {
...
-	coseSig := coseSignature{
-		Context:     "Signature1",
-		Protected:   cose.Protected,
-		ExternalAAD: []byte{},
-		Payload:     cose.Payload,
+		return Document{}, fmt.Errorf("attestation was generated in debug mode")
 	}
 
-	sigStruct, err := cbor.Marshal(&coseSig)
+	certificates, err := doc.Verify(certProvider, verificationTime)
```
Comment (mwittie) [inline internal/nitrite.go:74]: I love how `internal.Verify` came out - very clean!

File: internal/cose.go [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+		return fmt.Errorf("verifying signature: %w", err)
+	}
+	return nil
+}
+
+func (c CoseSign1) VerifySignature(
+	publicKey *ecdsa.PublicKey,
+) error {
+	if publicKey == nil {
+		return fmt.Errorf("public key is nil")
```
Comment (mwittie) [inline internal/cose.go:60]: I think we're missing a test case for nil public key

Comment (mwittie) [reply internal/cose.go:60]: Alternatively, you could pass in non-pointer `ecdsa.Public` key and avoid this check.

File: internal/cose.go [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+				"cose signing alg '%v' does not match public key signing alg '%s'",
+				signingAlg,
+				publicKey.Curve.Params().Name,
+			)
+		}
+
+		h := sha512.Sum384(sigStructBytes)
+		sigStructHash = h[:]
+	default:
+		return fmt.Errorf("unsupported signing alg '%v'", signingAlg)
```
Comment (mwittie) [inline internal/cose.go:102]: I think we're missing a test for this case.

File: internal/cose.go [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+			fmt.Errorf("unmarshaling cose protected header: %w", err)
+	}
+
+	intAlg, ok := header.AlgorithmInt()
+	if ok {
+		switch intAlg {
+		case -35:
+			return ES384, nil
+		default:
+			return BadOrMissingAlgorithm,
```
Comment (mwittie) [inline internal/cose.go]: In this case, I don't think we're missing an algorithm, since we're after `ok`. Maybe you could expand the `iota` with another sentinel value for `MissingSigningAlgorithm`?

File: internal/cose.go [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+		}
+	}
+
+	strAlg, ok := header.AlgorithmString()
+	if ok {
+		switch strAlg {
+		case "ES384":
+			return ES384, nil
+		default:
+			return BadOrMissingAlgorithm,
```
Comment (mwittie) [inline internal/cose.go]: Ditto

File: internal/cose.go [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+				fmt.Errorf("unsupported signing algorithm string '%s'", strAlg)
+		}
+	}
+	return BadOrMissingAlgorithm, fmt.Errorf("unsupported signing algorithm type")
```
Comment (mwittie) [inline internal/cose.go]: Here we are clearly missing the algorithm, so maybe `MissingSigningAlgorithm`?

**Review by mwittie — APPROVED** [review]
Looks really good! I was able to go through stuff with a fine comb and added a few very low level comments. :shipit:

File: internal/cose.go (hunk 23 above)
Comment (tahardi) [reply internal/cose.go]: Split up into two different enums: BadSigning Algorithm and MissingSigningAlgorithm

File: internal/cose.go (hunk 22 above)
Comment (tahardi) [reply internal/cose.go:102]: It's not testable under the current implementation as `GetSigningAlgorithm()` catches unsupported signing algos and is called earlier in `VerifySignature()`.
